### PR TITLE
Pinot batch ingestion hadoop

### PR DESCRIPTION
**Pull request by fx19880617**
- refactor common codes
- adding support to run pinot batch ingestion job in hadoop
- package plugins directory and untar them in hadoop mapper/ spark executor.
- Change PluginManager to class loader to not use system class loader

### CODE REVIEW

File: pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+import org.apache.pinot.spi.plugin.PluginManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.yaml.snakeyaml.Yaml;
+
+import static org.apache.pinot.plugin.ingestion.batch.common.SegmentGenerationUtils.PINOT_PLUGINS_TAR_GZ;
+import static org.apache.pinot.spi.plugin.PluginManager.PLUGINS_INCLUDE_PROPERTY_NAME;
+
+
+public class HadoopSegmentGenerationJobRunner extends Configured implements IngestionJobRunner, Serializable {
```
Comment (kishoreg) [inline pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java]: Add the logic here and some explanation on why we are compressing the plugins directory and shipping it to mapper via distributed cache.

File: pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+  /**
+   * Can be overridden to plug in custom mapper.
+   */
+  protected Class<? extends Mapper<LongWritable, Text, LongWritable, Text>> getMapperClass() {
+    return HadoopSegmentCreationMapper.class;
+  }
+
+  protected void packPluginsToDistributedCache(Job job) {
+    String pluginsRootDir = PluginManager.get().getPluginsRootDir();
+    File pluginsTarGzFile = new File(PINOT_PLUGINS_TAR_GZ);
```
Comment (kishoreg) [inline pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java]: check for the existence of this directory.

File: pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/pom.xml [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-jar-plugin</artifactId>
+        <version>2.5</version>
+        <configuration>
+          <forceCreation>true</forceCreation>
+        </configuration>
+      </plugin>
+      <plugin>
```
[the remark this answers is not in the document]
Comment (fx19880617) [reply pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/pom.xml]: true, will update all poms

File: pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/pom.xml [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+      <artifactId>jackson-databind</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>com.fasterxml.jackson.core</groupId>
+      <artifactId>jackson-annotations</artifactId>
+    </dependency>
+  </dependencies>
+
+  <build>
+    <plugins>
```
[the remark this answers is not in the document]
Comment (fx19880617) [reply pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/pom.xml]: will remove

**Pull request merged**